src/TourCMS.php

```diff
@@ -22,385 +22,537 @@
+        // Callback function to populate the response headers on curl_exec
+        $apiResponseHeaders = [];
+        curl_setopt(
+            $ch,
+            CURLOPT_HEADERFUNCTION,
+            function ($ch, $header) use (&$apiResponseHeaders) {
+                $len = strlen($header);
+                $header = explode(':', $header, 2);
+                if (count($header) < 2) // ignore invalid headers
+                    return $len;
+
+                $name = strtolower(trim($header[0]));
+                $apiResponseHeaders[$name] = trim($header[1]);
+
+                return $len;
+            }
+        );
+
+        $response = curl_exec($ch);
+
+        $this->lastResponseHeaders = $apiResponseHeaders;
+
+        $headerSize = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
+        $result = substr($response, $headerSize);
+
+        // Check whether we need to return raw XML or
+        // convert to SimpleXML first
+        if ($this->resultType == self::RESULT_TYPE_XML)
+            $result = simplexml_load_string($result);
+
+        $this->lastRequestHeaders = $this->headers;
+        $this->headers = [];
+
+        return $result;
+    }
+
+    /**
+     * get_base_url
+     *
+     * @author Paul Slugocki
+     * @return string
+     */
+    public function get_base_url(): string
+    {
+        return $this->baseUrl;
+    }
+
+    /**
+     * set_base_url
+     *
+     * @author Paul Slugocki
+     * @param $url New base url
+     * @return bool
+     */
+    public function set_base_url(string $url): void
+    {
+        $this->baseUrl = $url;
+    }
+
+    /**
+     * set_user_agent
+     *
+     * @author Francisco Martinez Ramos
+     * @return bool
+     */
+    public function set_user_agent(string $user_agent, bool $prepend = true): void
+    {
+        $this->prependCallerToUserAgent = $prepend;
+        $this->userAgent = $user_agent;
+    }
+
+    /**
+     * add_header
+     *
+     * @author Francisco Martínez Ramos
+     * @param string $header Key of the header
+     * @param string $value Value of the header
+     * @return void
+     */
+    public function add_header(string $header, string $value, bool $permanent = false): void
+    {
+        $newHeader = "$header: $value";
+
+        if ($permanent) {
+            array_push($this->permanentHeaders, $newHeader);
+        } else {
+            array_push($this->headers, $newHeader);
+        }
+    }
+
+    /**
+     * set_request_identifier
+     *
+     * @author Francisco Martínez Ramos
+     * @param string $value Value of the request identifier header
+     * @return void
+     */
+    public function set_request_identifier(string $value): void
+    {
+        $this->add_header(self::HEADER_X_REQUEST_ID, $value);
+    }
+
+    # Get last request headers
+
+    public function get_last_request_headers(): array
+    {
+        return $this->lastRequestHeaders;
+    }
+
+    # Get last response headers
+
+    public function get_lastResponseHeaders(): array
```

Comment:
this function name should have not changed, being public could be some customer using it